### ISSUE TITLE
Run evolved agent training on remote GPU sandbox

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#82]
## Objective

Advance our evolved agent research by running the training process on a remote GPU sandbox to enable faster iteration and longer training runs.

## Background

We've been developing GPU-accelerated training with PufferLib and PyTorch DDP. To properly validate and advance this research, we need to run extended training sessions on dedicated GPU hardware rather than local machines.

## Proposed Approach

1. **Remote Environment Setup**
   - Configure the remote GPU sandbox (rwalters-sandbox-1) for training
   - Ensure all dependencies (CUDA, PyTorch, PufferLib, etc.) are installed
   - Set up SSH access and port forwarding for monitoring

2. **Code Transfer & Configuration**
   - Transfer current training code to remote environment
   - Adapt paths and configurations for remote execution
   - Set up logging and checkpointing for long-running jobs

3. **Training Execution**
   - Run multi-GPU distributed training sessions
   - Monitor training progress remotely
   - Collect metrics and checkpoints

4. **Results Analysis**
   - Transfer trained models back for analysis
   - Compare performance across different configurations
   - Document findings for future iterations

## Acceptance Criteria

- [ ] Remote GPU sandbox is configured with all necessary dependencies
- [ ] Training code runs successfully on remote hardware
- [ ] Multi-GPU distributed training is functional
- [ ] Training metrics and checkpoints are being collected
- [ ] Can monitor training progress remotely
- [ ] Successfully complete at least one extended training run
- [ ] Document setup process and any remote-specific configurations

## Technical Notes

- Remote host: rwalters-sandbox-1 (SSH on port 22)
- MCP server already configured for remote access
- Need to handle file transfer for code and models
- Consider tmux/screen for persistent training sessions
- May need to adapt logging for headless operation

## Related Work

- Issue #8: GPU-accelerated training implementation
- Recent commits on multi-GPU DDP and PufferLib integration

**Comment by rjwalters** [comment]
## Implementation Guidance

### Phase 1: Remote Environment Setup

**Dependencies Check**:
```bash
# Verify CUDA availability
mcp__remote-ssh__remote_bash "nvidia-smi"

# Check Python version
mcp__remote-ssh__remote_bash "python3 --version"

# Verify package manager (uv preferred)
mcp__remote-ssh__remote_bash "which uv || echo 'Need to install uv'"
```

**Installation Steps**:
1. **Create remote workspace**:
   ```bash
   mcp__remote-ssh__remote_bash "mkdir -p ~/bucket-brigade-training && cd ~/bucket-brigade-training"
   ```

2. **Transfer codebase** using git:
   ```bash
   mcp__remote-ssh__remote_bash "cd ~/bucket-brigade-training && git clone https://github.com/rjwalters/bucket-brigade.git ."
   ```

3. **Setup Python environment**:
   ```bash
   mcp__remote-ssh__remote_bash "cd ~/bucket-brigade-training && uv venv && source .venv/bin/activate && uv pip install -e '.[rl]'"
   ```

4. **Verify GPU setup**:
   ```bash
   mcp__remote-ssh__remote_bash "cd ~/bucket-brigade-training && source .venv/bin/activate && python -c 'import torch; print(f\"CUDA available: {torch.cuda.is_available()}, GPUs: {torch.cuda.device_count()}\"')'"
   ```

### Phase 2: Training Configuration

**Key Files to Transfer/Configure**:
- `scripts/train_puffer_gpu.py` - Single GPU training (already exists)
- `scripts/train_multi_gpu.py` - Multi-GPU DDP training (already exists)
- Training configs and hyperparameters

**Remote-Specific Adaptations**:
1. **Persistent Sessions**: Use `tmux` or `screen` for long-running jobs
   ```bash
   mcp__remote-ssh__remote_bash "tmux new-session -d -s training 'cd ~/bucket-brigade-training && source .venv/bin/activate && python scripts/train_multi_gpu.py --num-steps 10000000'"
   ```

2. **Logging Configuration**: 
   - Ensure logs write to disk, not just console
   - Configure TensorBoard logs for remote viewing
   - Set up checkpoint directory with timestamps

3. **Resource Monitoring**:
   - Add GPU memory tracking
   - Monitor disk space for checkpoints
   - Track training throughput (steps/sec)

### Phase 3: Execution & Monitoring

**Launch Training**:
```bash
# For multi-GPU (4 GPUs example):
torchrun --nproc_per_node=4 scripts/train_multi_gpu.py --num-steps 10000000 --checkpoint-interval 50000
```

**Monitor Progress**:
```bash
# Check training logs
mcp__remote-ssh__remote_bash "tail -f ~/bucket-brigade-training/logs/training.log"

# Check GPU utilization
mcp__remote-ssh__remote_bash "watch -n 2 nvidia-smi"

# View TensorBoard metrics (if port forwarding configured)
# ssh -L 6006:localhost:6006 rwalters-sandbox-1
# tensorboard --logdir=~/bucket-brigade-training/runs
```

**Checkpoint Management**:
- Save checkpoints every 50k steps
- Keep last 3 checkpoints to manage disk space
- Transfer final model back to local machine

### Phase 4: Results Transfer & Analysis

**Download Trained Models**:
```bash
# Using scp or MCP file read
scp rwalters-sandbox-1:~/bucket-brigade-training/checkpoints/final_model.pt ./models/remote_trained_model.pt
```

**Metrics Collection**:
- Training curves (reward, loss, entropy)
- Throughput metrics (steps/sec, episodes/hour)
- GPU utilization stats
- Comparison with local training baseline

### Technical Considerations

**Existing Infrastructure**:
- ✅ MCP remote-ssh server configured in `.claude/settings.local.json`
- ✅ Remote host: `rwalters-sandbox-1` on port 22
- ✅ Training scripts already support multi-GPU: `scripts/train_multi_gpu.py`
- ✅ PufferLib integration in place: `scripts/train_puffer_gpu.py`
- ✅ Dependencies defined in `pyproject.toml` under `[project.optional-dependencies.rl]`

**Potential Issues & Solutions**:
1. **OOM (Out of Memory)**: Reduce batch size or number of parallel environments
2. **Slow I/O**: Use local SSD for checkpoints, transfer to persistent storage periodically
3. **Network instability**: Use tmux/screen to prevent training interruption on SSH disconnect
4. **Version mismatches**: Pin exact versions in `pyproject.toml` for reproducibility

### Success Metrics

- **Performance**: Training throughput ≥ 1000 steps/sec (vs ~100 local)
- **Stability**: Run completes without crashes for 10M+ steps
- **Quality**: Trained agent achieves competitive ELO rating
- **Reproducibility**: Can reproduce results with saved hyperparameters

### Implementation Phases

**Suggested Order**:
1. ✅ Start with single-GPU training to validate environment (lower risk)
2. ✅ Scale to multi-GPU once single-GPU works
3. ✅ Add monitoring and checkpointing
4. ✅ Run full training session (10M+ steps)
5. ✅ Transfer and analyze results

### Related Code References

- Training script: `scripts/train_puffer_gpu.py:1`
- Multi-GPU wrapper: `scripts/train_multi_gpu.py:1`
- PufferLib environment: `bucket_brigade/envs/puffer_env.py` (likely)
- Dependencies: `pyproject.toml:34-41`
